core/src/main/java/com/exadel/aem/toolkit/core/injectors/BaseInjectorTemplateMethod.java

```diff
@@ -0,0 +1,74 @@
+package com.exadel.aem.toolkit.core.injectors;
+
+import org.apache.sling.api.adapter.Adaptable;
+import org.apache.sling.models.spi.DisposalCallbackRegistry;
+import org.apache.sling.models.spi.Injector;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Type;
+import java.util.function.Supplier;
+
+public abstract class BaseInjectorTemplateMethod<T extends Annotation, R extends Adaptable> implements Injector {
+
+
+    protected T type;
+    protected R resource;
+
+    @CheckForNull
+    @Override
+    public Object getValue(
+        @Nonnull Object o,
+        String name,
+        @Nonnull Type type,
+        @Nonnull AnnotatedElement annotatedElement,
+        @Nonnull DisposalCallbackRegistry disposalCallbackRegistry
+    ) {
+        T annotation = getAnnotation(annotatedElement);
+
+        if (annotation == null) {
+            return null;
+        }
+
+        R adaptable = getAdaptable(o);
+
+        if(adaptable == null) {
+            return null;
+        }
+
+        Supplier<Object> annotationValueSupplier = getAnnotationValueSupplier(
+                                                adaptable,
+                                                name,
+                                                type,
+                                                annotatedElement,
+                                                disposalCallbackRegistry,
+                                                annotation );
+
+        Object value = annotationValueSupplier.get();
+
+        if( value == null) {
+            defaultMessage();
+        }
+
+        return value;
+    }
+
+    public abstract T getAnnotation(AnnotatedElement element);

```

Comment:
Consider lowering the visibility. TBD: are we going to make this abstract class a part of he Toolkit's public API (says: "everyone, you are now welcome to inherit your injectors from this!" . Seems unlikely ...) For the injectors of our very project, package private is enough

Reply:
corrected. Thank you.